rootex/framework/systems/render_system.cpp

```diff
@@ -100,9 +107,88 @@ void RenderSystem::popUIMatrix()
 	m_UITransformationStack.pop_back();
 }
 
+void RenderSystem::setProjectionConstantBuffers()
+{
+	const Matrix& projection = getCamera()->getProjectionMatrix();
+	if (m_VSProjectionConstantBuffer == nullptr)
+	{
+		D3D11_BUFFER_DESC cbd = { 0 };
+		cbd.BindFlags = D3D11_BIND_CONSTANT_BUFFER;
+		cbd.Usage = D3D11_USAGE_DYNAMIC;
+		cbd.CPUAccessFlags = D3D11_CPU_ACCESS_WRITE;
+		cbd.MiscFlags = 0u;
+		cbd.ByteWidth = sizeof(projection);
+		cbd.StructureByteStride = 0u;
+		D3D11_SUBRESOURCE_DATA csd = { 0 };
+		csd.pSysMem = &projection.Transpose();
+
+		m_VSProjectionConstantBuffer = RenderingDevice::GetSingleton()->createVSProjectionConstantBuffer(&cbd, &csd);
+	}
+	else
+	{
+		D3D11_MAPPED_SUBRESOURCE subresource = { 0 };
+		RenderingDevice::GetSingleton()->getBufferMappedContext(m_VSProjectionConstantBuffer.Get(), subresource);
+		memcpy(subresource.pData, &projection.Transpose(), sizeof(projection));
+		RenderingDevice::GetSingleton()->unmapBuffer(m_VSProjectionConstantBuffer.Get());
+	}
+}
+
+void RenderSystem::perFrameVSCBBinds()
+{
+	const Matrix& view = getCamera()->getViewMatrix();
+	if (m_VSPerFrameConstantBuffer == nullptr)
+	{
+		D3D11_BUFFER_DESC cbd = { 0 };
+		cbd.BindFlags = D3D11_BIND_CONSTANT_BUFFER;
+		cbd.Usage = D3D11_USAGE_DYNAMIC;
+		cbd.CPUAccessFlags = D3D11_CPU_ACCESS_WRITE;
+		cbd.MiscFlags = 0u;
+		cbd.ByteWidth = sizeof(view);
+		cbd.StructureByteStride = 0u;
+		D3D11_SUBRESOURCE_DATA csd = { 0 };
+		csd.pSysMem = &view.Transpose();
+
+		m_VSPerFrameConstantBuffer = RenderingDevice::GetSingleton()->createVSPerFrameConstantBuffer(&cbd, &csd);
+	}
+	else
+	{
+		D3D11_MAPPED_SUBRESOURCE subresource = { 0 };
+		RenderingDevice::GetSingleton()->getBufferMappedContext(m_VSPerFrameConstantBuffer.Get(), subresource);
+		memcpy(subresource.pData, &view.Transpose(), sizeof(view));
+		RenderingDevice::GetSingleton()->unmapBuffer(m_VSPerFrameConstantBuffer.Get());
+	}
+}
+
+void RenderSystem::perFramePSCBBinds()
+{
+	const Vector4& view = Vector4(getCamera()->getViewMatrix().Translation());
+	if (m_PSPerFrameConstantBuffer == nullptr)
+	{
+		D3D11_BUFFER_DESC cbd = { 0 };
+		cbd.BindFlags = D3D11_BIND_CONSTANT_BUFFER;
+		cbd.Usage = D3D11_USAGE_DYNAMIC;
+		cbd.CPUAccessFlags = D3D11_CPU_ACCESS_WRITE;
+		cbd.MiscFlags = 0u;
+		cbd.ByteWidth = sizeof(view);
+		cbd.StructureByteStride = 0u;
+		D3D11_SUBRESOURCE_DATA csd = { 0 };
+		csd.pSysMem = &view;
+
+		m_PSPerFrameConstantBuffer = RenderingDevice::GetSingleton()->createPSPerFrameConstantBuffer(&cbd, &csd);
+	}
+	else
+	{
+		D3D11_MAPPED_SUBRESOURCE subresource = { 0 };
+		RenderingDevice::GetSingleton()->getBufferMappedContext(m_PSPerFrameConstantBuffer.Get(), subresource);
+		memcpy(subresource.pData, &view, sizeof(view));
+		RenderingDevice::GetSingleton()->unmapBuffer(m_PSPerFrameConstantBuffer.Get());
+	}
+}
+
 void RenderSystem::setCamera(CameraComponent* camera)
 {
 	m_Camera = camera;
+	setProjectionConstantBuffers();
```

Comment:
Change `RenderSystem::restoreCamera()` to call `setCamera()` instead of touching the camera directly
https://github.com/sdslabs/Rootex/pull/158/files#diff-8a12acb9c977bc6c542f878e40109c3eR196